### ISSUE TITLE
ChatGLM3 get_mask() function ignore the length of prefix when using PEFT with P-tuningV2

### ISSUE THREAD

**Issue by ttw1018** [THUDM/ChatGLM3#359]
```python
batch_size, seq_length = input_ids.shape
full_attention_mask = torch.ones(batch_size, seq_length, seq_length, device=input_ids.device)
```

The correct mask length is len(input_ids) + len(prefix)

```python

full_attention_mask = torch.ones(batch_size, seq_length+len(prefix), seq_length+len(prefix), device=input_ids.device)
```

**Comment by zRzRzRzRzRzRzR** [comment]
@Btlmd 有空帮忙看一下

**Comment by ttw1018** [comment]
the PEFT get prompt is 
```python
past_key_values = past_key_values.view(
    batch_size,
    peft_config.num_virtual_tokens,
    peft_config.num_layers * 2,
    peft_config.num_attention_heads,
    peft_config.token_dim // peft_config.num_attention_heads,
)
past_key_values = past_key_values.permute([2, 0, 3, 1, 4]).split(
    peft_config.num_transformer_submodules * 2
)
```

but the official code of ChatGLM3 is 
```python
past_key_values = past_key_values.view(
    batch_size,
    self.pre_seq_len,
    self.num_layers * 2,
    self.multi_query_group_num,
    self.kv_channels
)
# seq_len, b, nh, hidden_size
past_key_values = self.dropout(past_key_values)
past_key_values = past_key_values.permute([2, 1, 0, 3, 4]).split(2)
```

Does it mean that chatglm3 can't be tuned with PEFT by prefix tuning?

**Comment by Btlmd** [comment]
I'm not sure if ChatGLM3 is compatible with PEFT. This hasn't been tested.

---


> ```python
>         batch_size, seq_length = input_ids.shape
>         full_attention_mask = torch.ones(batch_size, seq_length, seq_length, device=input_ids.device)
> ```
> 
> The correct mask length is len(input_ids) + len(prefix)
> 
> ```python
>     full_attention_mask = torch.ones(batch_size, seq_length+len(prefix), seq_length+len(prefix), device=input_ids.device)
> ```

Line 684-686

```Python
        if past_length:
            full_attention_mask = torch.cat((torch.ones(batch_size, seq_length, past_length,
                                                        device=input_ids.device), full_attention_mask), dim=-1)
```

will add the attention mask for the prefix if necessary.

**Comment by ttw1018** [comment]
> the PEFT get prompt is
> 
> ```python
> past_key_values = past_key_values.view(
>     batch_size,
>     peft_config.num_virtual_tokens,
>     peft_config.num_layers * 2,
>     peft_config.num_attention_heads,
>     peft_config.token_dim // peft_config.num_attention_heads,
> )
> past_key_values = past_key_values.permute([2, 0, 3, 1, 4]).split(
>     peft_config.num_transformer_submodules * 2
> )
> ```
> 
> but the official code of ChatGLM3 is
> 
> ```python
> past_key_values = past_key_values.view(
>     batch_size,
>     self.pre_seq_len,
>     self.num_layers * 2,
>     self.multi_query_group_num,
>     self.kv_channels
> )
> # seq_len, b, nh, hidden_size
> past_key_values = self.dropout(past_key_values)
> past_key_values = past_key_values.permute([2, 1, 0, 3, 4]).split(2)
> ```
> 
> Does it mean that chatglm3 can't be tuned with PEFT by prefix tuning?

The reason for the error is that the shape of prefix vector is not in line with the dimension in PEFT.

**Comment by Btlmd** [comment]
If the shape of prefix vector is mismatched, ChatGLM3 probably cannot be fine-tuned in PEFT by prefix tuning directly. You may consider the P-tuning v2 implementation in this repo for parameter-efficient finetuning.

**Comment by zRzRzRzRzRzRzR** [comment]
If you have more fine-tuning issues, you can raise them in #253. This issue is closed. Thank you for your support.
